server/src/main/java/org/apache/druid/server/http/CatalogResource.java

```diff
@@ -0,0 +1,488 @@
+import java.util.List;
+
+/**
+ * REST endpoint for user and internal catalog actions. Catalog actions
+ * occur at the global level (all schemas), the schema level, or the
+ * table level.
+ *
+ * @see {@link CatalogListenerResource} for the client-side API.
+ */
+@Path(CatalogResource.ROOT_PATH)
+public class CatalogResource
+{
+  public static final String ROOT_PATH = "/druid/coordinator/v1/catalog";
+
+  private final CatalogStorage catalog;
+
+  @Inject
+  public CatalogResource(CatalogStorage catalog)
+  {
+    this.catalog = catalog;
+  }
+
+  /**
+   * Create a new table within the indicated schema.
+   *
+   * @param table The table specification to create.
+   * @param ifNew Whether to skip the action if the table already exists.
+   *        This is the same as the SQL IF NOT EXISTS clause. If {@code false},
+   *        then an error is raised if the table exists. If {@code true}, then
+   *        the action silently does nothing if the table exists. Primarily for
+   *        use in scripts.
+   * @param req the HTTP request used for authorization.
+   * @return the version number of the table
+   */
+  @POST
+  @Path("/tables")
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response createTable(
+      TableMetadata table,
+      @QueryParam("ifnew") boolean ifNew,
+      @Context final HttpServletRequest req)
+  {
+    String dbSchema = table.resolveDbSchema();
+    Pair<Response, SchemaSpec> result = validateSchema(dbSchema);
+    if (result.lhs != null) {
+      return result.lhs;
+    }
+    SchemaSpec schema = result.rhs;
+    if (!schema.writable()) {
+      return Actions.badRequest(
+          Actions.INVALID,
+          StringUtils.format("Cannot create tables in schema %s", dbSchema));
+    }
+    table = table.withSchema(dbSchema);
+    try {
+      table.validate();
+    }
+    catch (IAE e) {
+      return Actions.badRequest(Actions.INVALID, e.getMessage());
+    }
+    TableSpec spec = table.spec();
+    if (!schema.accepts(spec)) {
+      return Actions.badRequest(
+          Actions.INVALID,
+          StringUtils.format(
+              "Cannot create tables of type %s in schema %s",
+              spec == null ? "null" : spec.getClass().getSimpleName(),
+              dbSchema));
+    }
+    try {
+      catalog.authorizer().authorizeTable(schema, table.name(), Action.WRITE, req);
+    }
+    catch (ForbiddenException e) {
+      return Actions.forbidden(e);
+    }
+    try {
+      long createVersion = catalog.tables().create(table);
+      return Actions.okWithVersion(createVersion);
+    }
+    catch (DuplicateKeyException e) {
+      if (!ifNew) {
+        return Actions.badRequest(
+              Actions.DUPLICATE_ERROR,
+              StringUtils.format(
+                  "A table of name %s.%s aleady exists",
+                  table.dbSchema(),
+                  table.name()));
+      } else {
+        return Actions.okWithVersion(0);
```

Comment:
Should this return the current version of the already-existing table?

Reply:
It could, but that would take an extra DB read. This feature mimics the SQL `CREATE TABLE myTempTable IF NOT EXISTS` use case where all we want is to not fail if we've already done this step; we typically don't then change the definition.

If we do want to change anything, we've got to read the existing values, which would provide the version. By not providing the version here, we save a DB read internally, since the failed SQL `INSERT` didn't fetch it.